kiesraad/abacus
# Asset serving via `memory-serve`

Closes [#365](https://github.com/kiesraad/abacus/issues/365)

- Adds [memory-serve](https://github.com/tweedegolf/memory-serve) to serve the frontend assets.
- Adds `tracing` and `tracing-subscriber` for logging, replaced `println!` and `eprintln!` calls with `tracing` equivalents
- Removes serving via `tower`'s `ServeDir` and removes the dependencies
- Removes `frontend_dist` command line argument, as `memory-serve` needs to know the path at compile time.

To test: 
- run `cargo run` and see that only the API is served
- Make sure that the `ASSET_DIR` environment variable is set to the `frontend/dist` folder
- run `cargo run --features memory-serve` to see in the logs that /index.html is served as index and fallback. Also see that the frontend is available via `http://localhost:8080`.

In normal development builds, the assets are loaded ad-hoc from disk, so changes made to the files in `frontend/dist` will be reflected on a page refresh.

In release builds, the assets are embedded into the binary and served from memory. 

----
### ToDo:
When `memory-serve` has a new release
- [x] Document the usage of ASSET_DIR when still needed
- [x] Update playwright config to use ASSET_DIR
